apps/builder/app/canvas/features/webstudio-component/webstudio-component.tsx

```diff
@@ -119,6 +127,68 @@ const getInstanceSelector = (
   return undefined;
 };
 
+const $indexesWithinAncestors = computed(
+  [registeredComponentMetasStore, instancesStore, selectedPageStore],
+  (metas, instances, page) => {
+    return getIndexesWithinAncestors(
+      metas,
+      instances,
+      page ? [page.rootInstanceId] : []
+    );
+  }
+);
+
+export const WebstudioComponentContext = createContext<{
+  propsByInstanceIdStore: Atom<Map<Instance["id"], Prop[]>>;
+}>({
+  propsByInstanceIdStore: atom(new Map()),
+});
+
+const useInstanceProps = (instanceId: Instance["id"]) => {
+  const { propsByInstanceIdStore } = useContext(WebstudioComponentContext);
+  const instancePropsObjectStore = useMemo(() => {
+    return computed(
+      [propsByInstanceIdStore, dataSourcesLogicStore, $indexesWithinAncestors],
+      (propsByInstanceId, dataSourcesLogic, indexesWithinAncestors) => {
```

Comment:
this feels like something that should have some tests

Reply:
Might be refactored even more later

Reply:
jsut lots of logic here with no certainty it works